app/src/main/java/org/schabi/newpipe/player/mediabrowser/MediaBrowserConnector.kt

```diff
@@ -0,0 +1,814 @@
+                }
+                val playlistId = path.get(0).toLong()
+                val index = path.get(1).toInt()
+                return if (playlistType == ID_LOCAL)
+                    extractLocalPlayQueue(playlistId, index)
+                else
+                    extractRemotePlayQueue(playlistId, index)
+            }
+
+            ID_URL -> {
+                if (path.size != 1) {
+                    throw parseError(mediaId)
+                }
+
+                val serviceId = path.get(0).toInt()
+                return ExtractorHelper.getPlaylistInfo(serviceId, url, false)
+                    .map<PlayQueue>({ info: PlaylistInfo? -> PlaylistPlayQueue(info) })
+            }
+
+            else -> throw parseError(mediaId)
+        }
+    }
+
+    @Throws(ContentNotAvailableException::class)
+    private fun extractPlayQueueFromHistoryMediaId(
+        mediaId: String,
+        path: List<String>
+    ): Single<PlayQueue> {
+        if (path.size != 1) {
+            throw parseError(mediaId)
+        }
+
+        val streamId = path.get(0).toLong()
+        return database.streamHistoryDAO().getHistory()
+            .firstOrError()
+            .map<PlayQueue>(
+                Function { items: MutableList<StreamHistoryEntry?>? ->
+                    val infoItems = items!!.stream()
+                        .filter { it: StreamHistoryEntry? -> it!!.streamId == streamId }
+                        .map<StreamInfoItem?> { obj: StreamHistoryEntry? -> obj!!.toStreamInfoItem() }
+                        .collect(Collectors.toList())
+                    SinglePlayQueue(infoItems, 0)
+                }
+            )
+    }
+
+    override fun getSupportedPrepareActions(): Long {
+        return PlaybackStateCompat.ACTION_PLAY_FROM_MEDIA_ID
+    }
+
+    private fun disposePrepareOrPlayCommands() {
+        prepareOrPlayDisposable?.dispose()
+    }
+
+    override fun onPrepare(playWhenReady: Boolean) {
+        disposePrepareOrPlayCommands()
+        // No need to prepare
+    }
+
+    override fun onPrepareFromMediaId(
+        mediaId: String,
+        playWhenReady: Boolean,
+        extras: Bundle?
+    ) {
+        if (MainActivity.DEBUG) {
+            Log.d(
+                TAG,
+                String.format(
+                    "MediaBrowserConnector.onPrepareFromMediaId(%s, %s, %s)",
+                    mediaId, playWhenReady, extras
+                )
+            )
+        }
+
+        disposePrepareOrPlayCommands()
+        prepareOrPlayDisposable = extractPlayQueueFromMediaId(mediaId)!!
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(
+                { playQueue: PlayQueue? ->
+                    sessionConnector.setCustomErrorMessage(null)
+                    NavigationHelper.playOnBackgroundPlayer(
+                        context, playQueue,
+                        playWhenReady
+                    )
+                },
```

Comment:
Especially cause we use `NavigationHelper.playOnBackgroundPlayer` here in our `PlaybackPreparer` instance, which in turn starts the `Intent`, which in turn ensures that `PlayerService` runs, which then tries to set up the media session controller again.

We have completely lost any hold of our lifecycle management at this point.